File: python/komorph/scorer.py
from typing import List, Tuple
from .rules import MorphemeCandidate

class MorphemeScorer:
    def __init__(self):
        # 품사별 기본 점수
        self.pos_scores = {
            "N": 0.1,   # 명사
            "V": 0.2,   # 동사
            "A": 0.2,   # 형용사
            "J": 0.3,   # 조사
            "E": 0.3,   # 어미
        }
        
        # 품사 연접 점수 (bigram)
        self.transition_scores = {
            ("N", "J"): 0.5,    # 명사 + 조사
            ("V", "E"): 0.5,    # 동사 + 어미
            ("A", "E"): 0.5,    # 형용사 + 어미
        }
    
    def score_candidate(self, candidate: MorphemeCandidate) -> float:
        """후보에 대한 종합 점수 계산"""
        analysis = candidate.analysis
        score = candidate.score  # 기본 점수에서 시작
        
        # 1) 형태소 길이 보너스
        for morph, pos in analysis:
            if pos in ["N", "V", "A"]:  # 실질 형태소
                length_bonus = min(len(morph) * 0.05, 0.3)
                score += length_bonus
        
        # 2) 품사 연접 보너스
        if len(analysis) == 2:
            pos1, pos2 = analysis[0][1], analysis[1][1]
            if (pos1, pos2) in self.transition_scores:
                score += self.transition_scores[(pos1, pos2)]
        
        # 3) 너무 짧은 토큰 페널티
        total_length = sum(len(morph) for morph, _ in analysis)
        if total_length <= 1:
            score -= 0.2
        
        return score
    
    def rank_candidates(self, candidates: List[MorphemeCandidate]) -> List[MorphemeCandidate]:
        """후보들을 점수순으로 정렬"""
        for candidate in candidates:
            candidate.score = self.score_candidate(candidate)
        
        return sorted(candidates, key=lambda c: c.score, reverse=True)

File: rust/src/scorer.rs
use crate::rules::Cand;

pub fn score_candidate(analysis: &Cand) -> f32 {
    let mut score = 0.1;
    let tags: Vec<&str> = analysis.iter().map(|(_, t)| t.as_str()).collect();
    
    // 분해 보너스
    if tags.iter().any(|t| *t == "J" || *t == "E") {
        score += 0.5;
    }
    
    // 형태소 길이 보너스
    for (morph, pos) in analysis {
        if matches!(pos.as_str(), "N" | "V" | "A") && morph.chars().count() >= 2 {
            score += (morph.chars().count() as f32 * 0.03).min(0.2);
        }
    }
    
    // 품사 연접 보너스
    if analysis.len() == 2 {
        let (_, pos1) = &analysis[0];
        let (_, pos2) = &analysis[1];
        match (pos1.as_str(), pos2.as_str()) {
            ("N", "J") | ("V", "E") | ("A", "E") => score += 0.5,
            _ => {}
        }
    }
    
    // 은/는 모호성 처리
    if analysis.len() >= 2 {
        let (suf, tag) = &analysis[1];
        if matches!(suf.as_str(), "은" | "는") {
            if tag == "E" {
                score -= 0.4;
            } else if tag == "J" {
                score += 0.3;
            }
        }
    }
    
    // 너무 짧은 토큰 페널티
    let total_len: usize = analysis.iter().map(|(m, _)| m.chars().count()).sum();
    if total_len <= 1 {
        score -= 0.2;
    }
    
    score
}